IGNITE-28671 Describe healthy cluster behavior in general tips guide (#13130)
apache/ignite: docs/_docs/perf-and-troubleshooting/general-perf-tips.adoc

* Adjust link:data-rebalancing[data rebalancing settings] to ensure that rebalancing completes faster when your cluster topology changes.

> == What healthy cluster behavior looks like

Reviewer: Let's add examples for `control.sh` commands and JMX beans screenshot that demonstrates your article.

Like the following:

```
$ ./control.sh --cluster state
... example output for ACTIVE cluster state here ...

$ ./control.sh --baseline
... example output for healthy baseline when all nodes are active here ...
... and so on ...
```

For those checks that must be in graph form let's provide simplest jconsole examples.

Author: done

> A healthy Ignite cluster is not defined by a single latency, CPU, or memory number. In practice, it is a cluster whose topology is stable, whose cluster state and baseline match the intended deployment, whose partitions are not lost or divergent, whose rebalancing and checkpointing complete in bounded time, and whose execution queues and memory pools return to a steady level after short-lived spikes. Ignite exposes these signals through built-in metrics, system views, and the control script rather than through a single aggregate health score.

Reviewer 2: `A healthy Ignite cluster is not defined by a single latency, CPU, or memory number.` - very strange comment

Author: yea, thx. Looks wierd at  second glance, i agree. i was intended to say smth like: "healthy cluster could not bejust defined by some simple memory metrics (number), but rather... whole complex system etc"

> When checking whether a cluster is healthy, start with topology and cluster state. The cluster should be in the expected state, usually ACTIVE, and the number of server and client nodes should be stable. If native persistence is enabled, the baseline should also be in the expected shape: for a stable deployment, the nodes that are expected to be online should appear online both in baseline-related metrics and in the SYS.BASELINE_NODES system view. Frequent unexpected topology changes are not normal and should be treated as a sign of node instability or network problems.

Reviewer 2: usually ACTIVE - need cross link to ACTIVE

Author: thank you, links added

Checkpointing and transactions should also remain bounded. Checkpoint activity can slow the cluster down, so LastCheckpointDuration should be monitored together with dirty pages and disk behavior. Transactions and queries can legitimately take longer during bursts, but healthy steady-state behavior means that lock-holding transactions, long-running transactions, and long-running SQL queries do not accumulate over time. If long transactions repeatedly block partition map exchange, use transaction timeout settings such as TxTimeoutOnPartitionMapExchange and investigate the application path that keeps transactions open.

> Finally, check the underlying JVM and critical workers. Ignite treats IgniteOutOfMemoryException, OutOfMemoryError, system worker termination, system worker hangs, and cluster node segmentation as critical failures. A healthy cluster should not emit blocked system-critical worker messages, and JVM resource pools should stay comfortably below exhaustion. In practice, monitor heap usage, direct buffer usage, and open file descriptors continuously, because all three are finite pools and approaching their limits usually means the node is already close to a failure condition rather than merely under benign load.

Reviewer 2: `as critical failures` - and what does it mean ?

Author: Completely refurbish this phrase. Before that, i meant to think about critical failures as smth that triggers FailureHandler. Thank you

* Adjust link:data-rebalancing[data rebalancing settings] to ensure that rebalancing completes faster when your cluster topology changes.

== How to assess cluster health

Cluster health is a complex thing. Apache Ignite is capable of demonstrating great performance across different scenarios with varying loads. Therefore, in general terms, a healthy cluster is one whose behavior aligns with your expectations. However, there are some universal aspects that apply to all deployments and warrant attention.

It is important to understand that a healthy cluster may undergo planned topology changes or temporary load spikes.

The key properties are:
- The cluster serves operations allowed by its intended state.

> - Data remains consistent (`idle_verify` passes, meaning partitions are intact and there are no conflicts).

Reviewer 2: 'idle_verify' need to be cross-linked with appropriate chapter

Author: done

> - Expected nodes are present.

Reviewer 2: what does it mean in common ? Probably we need to mention baseline, partition loss ?

- No pending transactions and no split brain.
- No lost partitions.
- Transient activity such as rebalancing or queue growth converges instead of accumulating.

> But the thing is, there is no universal single health metric.

Reviewer 2: From reader side this phrase is confusing, (you and me understand that it\`s true but i don\`t like to show it for reader) we need to remove it at all or extend this comment.

Author: extended

Use several signals together.

> For example, `SELECT 1` can confirm that a client or SQL query path is reachable, but it does not check user partitions, backup consistency, baseline membership, or all server nodes.

Reviewer 2: I know that this is a common practise (in some installations) but, in common, it shows nothing, this command can be executed on client node without cluster nodes interaction.

Author: completely refurbished this line

> Similarly, `control.sh --cache idle_verify` is an important consistency check, but it is not a complete health check.

Reviewer 2: control.**sh** - it\`s a linux specific

Author: fixed overwhere, thx

=== Check the Intended State and Node Membership
Start with the link:monitoring-metrics/cluster-states[cluster state].

Run:

[source,shell]
----
./control.sh --state
----

Relevant output:

[source,text]
----
Command [STATE] started
Arguments: --state
--------------------------------------------------------------------------------
Cluster state: ACTIVE
Command [STATE] finished with code: 0
----

`ACTIVE` is expected for normal read-write operation.
`ACTIVE_READ_ONLY` is normal when read-only operation was intentionally enabled.
`INACTIVE` is acceptable only when it matches the current operation, for example planned maintenance; an inactive cluster does not serve the data workload.
The criterion is whether the actual state matches the state that was intentionally set for the deployment.

> If baseline topology is enabled, check the baseline as well:

Reviewer 2: When it can be enabled\disabled and what user need to do - if it  disabled ?

Run:

[source,shell]
----
./control.sh --baseline
----

Relevant output for a cluster where all baseline nodes are online:

[source,text]
----
Cluster state: ACTIVE
Current topology version: 3
Baseline auto adjustment disabled: softTimeout=300000

Current topology version: 3 (Coordinator: ConsistentId=node-1, Order=1)

Baseline nodes:
    ConsistentId=node-1, State=ONLINE, Order=1
    ConsistentId=node-2, State=ONLINE, Order=2
    ConsistentId=node-3, State=ONLINE, Order=3
--------------------------------------------------------------------------------
Number of baseline nodes: 3

Other nodes not found.
----

> If an online server node has joined the cluster but is not in the baseline, the command shows it under `Other nodes`:

Reviewer 2: Plz append examples: 
1. with one lost node
2. with partition loss
 2a. if node will come back to cluster - it still has a partition loss and need to be reset - expand this plz.

[source,text]
----
Other nodes:
    ConsistentId=node-4, Order=4
Number of other nodes: 1
----

The baseline contains server nodes that are intended to store data.
Client nodes are not part of the baseline.
An online server node in `Other nodes` is not always an error: the node may have been prepared intentionally but not yet introduced into the data topology.

> If the node is expected to store data, first check the link:clustering/baseline-topology#baseline-topology-autoadjustment[baseline auto-adjustment policy] and the deployment plan, then use the documented baseline change procedure.

Reviewer 2: "deployment plan" - what does it mean ?

Author: extended to fulfil this thought

> Changing the baseline can start link:data-rebalancing[rebalancing]. This is often not an easy procedure for a cluster with data in persistence storage.

Reviewer 2: "easy" - not an engeneering term ) what does it mean ? I\`d prefer to say like : Baseline change procedure will provide 'rebalance' which means that partitions will be rebased according to new affinity distribution.

Author: extended with condition-and-action structure. i think it is the clearest method of explanation

Frequent unexpected topology changes are not normal and usually indicate node instability, network issues, or misconfigured discovery. However, a temporary topology change does not make the cluster unhealthy by itself.

> Server `JOIN`, `LEFT`, and `FAIL` events change cluster membership and the major topology version.

Reviewer 2: Not clear, rewrite or expand it plz.

Author: done

> Cache start or stop, distributed cache creation, SQL table creation, and other affinity-changing events can change the minor affinity topology version without a node loss.

Reviewer 2: What different between: 'Cache start' and 'distributed cache creation' ?
'SQL table creation', and what about drop ?
And over all - why this information is helpful in acpect of this document ?
Probably you want to say smth., but i miss it (

Author: confusing stuff was replaced by 'Cache or SQL schema changes can also trigger PME without server node loss. Therefore, a topology version or PME metric change is useful only when interpreted together with maintenance actions and node logs.'

Planned events are normal.

> Investigate unexpected repeated membership churn, node segmentation, network failures, or a partition map exchange that does not finish.

Reviewer 2: Just give me a clue here - this sentence is correct, but how it can be detected ? I think we need to form this document from position that reader is not deeply understant how it work and need to have some commands\logs how to understand (split brain\PME) situations ?

Author: Commands added, could be expanded with logs anyway.. 

Should we expand split-brain situations? It is kinda vague and misty in this doc from my point of view now

* Adjust link:data-rebalancing[data rebalancing settings] to ensure that rebalancing completes faster when your cluster topology changes.

== How to assess cluster health

Cluster health is a complex thing. Apache Ignite is capable of demonstrating great performance across different scenarios with varying loads. Therefore, in general terms, a healthy cluster is one whose behavior aligns with your expectations. However, there are some universal aspects that apply to all deployments and warrant attention.

It is important to understand that a healthy cluster may undergo planned topology changes or temporary load spikes.

The key properties are:

* The cluster is in the intended link:monitoring-metrics/cluster-states[state] and serves only the operations allowed by that state.
* Baseline topology, when it is used or managed manually, matches the expected data-bearing server nodes.
* Data remains consistent: link:tools/control-script#verifying-partition-checksums[`idle_verify`] reports no partition conflicts when the cluster is idle.

> * Expected nodes are present, and no node segmentation or repeated membership churn is reported.

Reviewer 2: "repeated membership churn" - i don\`t listen this term before, what does it mean ?

Author: churn-change describes sitations that happens in the set of participating nodes participating nodes due to joins, graceful leaves, and failures. For example, a high churn rate can increase costs or decrease service quality. [src](https://www.academia.edu/100244898/Minimizing_churn_in_distributed_systems): 

as for this article, i'll replace it with more common wording

* link:configuring-caches/partition-loss-policy[Lost partitions] are absent.

> * Long-running transactions, Partition Map Exchange (PME), rebalancing, checkpointing, and executor queues converge instead of accumulating.

Reviewer 2: All this terms need to be covered\documentally extended somehow ? PME - what does it mean ? Long-running transactions - how it can be detected ? and so on ...

Author: there will be detailed explanation of this metrics later on in arictle, but here is the first use of it in text. so i've linked it to their doc origins

There is no single command or metric that proves cluster health for every deployment.

Use several signals together.
A simple client connection or SQL liveness check can prove only that a particular client or query path is reachable; it does not check user partitions, backup consistency, baseline membership, or all server nodes.

> Similarly, `control.(sh|bat) --cache idle_verify` is an important consistency check, but it is still not a complete health check.

Reviewer 2: I misunderstand this section at all ( It gives nothing only confusing a reader, need to be refactored or removed.

Author: last line removed. finally

Author: everything removed now

Baseline nodes:
    ConsistentId=node-1, State=ONLINE, Order=1
    ConsistentId=node-2, State=ONLINE, Order=2
    ConsistentId=node-3, State=ONLINE, Order=3
--------------------------------------------------------------------------------
Number of baseline nodes: 3

Other nodes not found.
----

Example: one baseline node is offline:

[source,text]
----
Baseline nodes:
    ConsistentId=node-1, State=ONLINE, Order=1
    ConsistentId=node-2, State=OFFLINE, Order=2
    ConsistentId=node-3, State=ONLINE, Order=3
--------------------------------------------------------------------------------
Number of baseline nodes: 3
----

If a baseline node is `OFFLINE`, an expected data-bearing server is missing.
If other primary or backup copies are available, its absence does not cause partition loss.
To check for partition loss, query the partition states as described in <<confirm-that-rebalancing-converges,Confirm That Rebalancing Converges>>.

If an online server node has joined the cluster but is not in the baseline, the command shows it under `Other nodes`:

[source,text]
----
Other nodes:
    ConsistentId=node-4, Order=4
Number of other nodes: 1
----

> The baseline contains server nodes that are intended to store data.

Reviewer 2: ```suggestion
The baseline contains server nodes that are intended to store a data.
```

> Client nodes are not part of the baseline.

Reviewer 2: ```suggestion
Client nodes are not a part of the baseline.
```

An online server node in `Other nodes` is not always an error: the node may have been prepared intentionally but not yet introduced into the data topology.
If the node is expected to store data, first check the link:clustering/baseline-topology#baseline-topology-autoadjustment[baseline auto-adjustment policy] and the current maintenance or scale-out procedure, then use the documented baseline change procedure.
Changing the baseline can start link:data-rebalancing[rebalancing]: partitions are redistributed according to the new affinity assignment.
Plan for the additional network, CPU, and storage load, especially in clusters with persistence.

Use topology changes to distinguish planned activity from instability.

> A server `JOIN`, `LEFT`, or `FAIL` event changes cluster membership and triggers PME.

Reviewer 2: PME - crosslink or extended description need to be passed

Author: thx

> Cache or SQL schema changes can also trigger PME without server node loss.

Reviewer 2: I misunderstand here, what does it mean - " without server node loss"  ? May be i miss smth ? plz explain ?

Author: i wanted to say that not only so called churn-change can trigger PME, but schema changes as well like cache start\stop. The user may want to know that there is no need to panic, it is not neccessary server loss or smth

> Therefore, a topology version or PME metric change is useful only when interpreted together with maintenance actions and node logs.

Reviewer 2: PME metric - crosslink ? how to check this metric ?

Author: crosslinked

Run `control.(sh|bat) --baseline` repeatedly or monitor topology metrics to confirm that membership is stable when no maintenance is in progress.
In logs, look for repeated node join, left, fail, segmentation, and exchange-worker messages.

> Investigate unexpected repeated membership churn, node segmentation, network failures, or a PME that does not finish.

Reviewer 2: 'churn' - confusing, may be only for me ?

Author: i agree. it's not common use at all. From the "newbie hat" perspective it makes no sense

=== Verify Partition Consistency

When the cluster is expected to be idle, run:

[source,shell]
----
control.(sh|bat) --cache idle_verify
----

Successful result:

[source,text]
----
The check procedure has finished, no conflicts have been found.
----

The beginning of a conflict result uses this format:

[source,text]
----
The check procedure has failed, conflict partitions has been found: [counterConflicts=1, hashConflicts=0]
Update counter conflicts:
Conflict partition: PartitionKey [grpId=1544803905, grpName=default, partId=5]
----

The command compares partition update counters and partition hashes between primary and backup copies.
Run it only when data updates are stopped.
If updates are active, the command can report false conflicts because copies are changing while hashes are being calculated.
Partitions in `MOVING` or `LOST` state may be skipped, so the result can be incomplete.
A successful `idle_verify` result is an important confirmation of consistency, but it still does not prove overall cluster health check success.

[#confirm-that-rebalancing-converges]
=== Confirm That Rebalancing Converges

> After a topology event, such as a new node join, transient rebalancing is expected, but it should converge. Rebalance progress should move toward completion.

Reviewer 2: ```suggestion
After a topology event raized, such as: a new node join\left, ???transient??? rebalancing is expected, but it should converge. Rebalance progress should move toward completion.
```

"converge"  - gives nothing, probably you need to show some metric examples ?

Author: fixed

[source,sql]
----
SELECT CACHE_GROUP_ID, PARTITION_ID, NODE_ID, STATE, IS_PRIMARY
FROM SYS.PARTITION_STATES
WHERE STATE IN ('MOVING', 'RENTING', 'LOST')
ORDER BY STATE, CACHE_GROUP_ID, PARTITION_ID, NODE_ID;
----

In steady state, this query usually should not return `MOVING`, `RENTING`, or `LOST` rows.
`MOVING` and `RENTING` are expected right after an intended topology or cache event, but their count should decrease.
`LOST` is not a normal transient state.

Example: one partition is lost:

[source,text]
----
CACHE_GROUP_ID | PARTITION_ID | NODE_ID                              | STATE | IS_PRIMARY
1544803905     | 5            | 0f4d6f30-3e04-4f68-b6a2-6b89f1795c0d | LOST  | true
----

If the query returns `LOST`, follow the link:configuring-caches/partition-loss-policy[Partition Loss Policy] recovery procedure.
If a failed node returns, its persistent data may become available again, but the affected partitions remain in the `LOST` state.
If the required data is available again, reset the lost partitions:

[source,shell]
----
control.(sh|bat) --cache reset_lost_partitions cacheName1,cacheName2,...
----

> If data is still missing, resetting the partitions does not restore it; the command only clears the `LOST` state.

Reviewer 2: Very confusing sentence and very scary, as for me. Plz explain - "If data is still missing" how it can happen ? 
for example backup = 2 node1 - left the grid, node2 (with second backup) is still in topology, no autoadjust is happens, some writes are going into node2, further - lost node2 and return node1 = miss data, smth like this

Author: added on with some explanarory examples

On an idle node, queues usually return to zero.
Investigate continuous growth, lack of drain after load stops, repeated `DetectStarvation=true`, or repeated starvation warnings in logs.
There is no universal absolute threshold.
Queue metrics are node-local, so collect them from all server nodes.

JMX uses the metric registry name to build `group` and `name` in the MBean object name.
The following mappings are useful for queue checks:

* Registry `io.communication` is exposed as JMX group `io`, bean name `communication`; the attribute is `OutboundMessagesQueueSize`.
* Registry `io.discovery` is exposed as JMX group `io`, bean name `discovery`; the attribute is `MessageWorkerQueueSize`.
* Registry `threadPools.StripedExecutor` is exposed as JMX group `threadPools`, bean name `StripedExecutor`; the attributes include `TotalQueueSize`, `StripesQueueSizes`, and `DetectStarvation`.
* Regular pools such as `threadPools.GridSystemExecutor` expose `QueueSize`.

For JMX object names and SQL metric access, see link:monitoring-metrics/new-metrics-system#jmx[JMX] and link:monitoring-metrics/new-metrics-system#sql-view[SQL View].

> .JConsole view of node-local striped executor queue metrics

Reviewer 2: ```suggestion
JConsole view of node-local striped executor queue metrics
```

Author: wont fix, ascii formatting issue

If the checkpoint buffer is exhausted, update processing can stop until the checkpoint completes.

Monitor these metrics for persistent data regions and data storage:

* `io.dataregion.<region>.DirtyPages`
* `io.dataregion.<region>.CheckpointBufferSize`
* `io.dataregion.<region>.UsedCheckpointBufferSize`
* `io.dataregion.<region>.TotalThrottlingTime`
* `io.datastorage.LastCheckpointStart`
* `io.datastorage.LastCheckpointDuration`
* `io.datastorage.LastCheckpointPagesWriteDuration`
* `io.datastorage.LastCheckpointTotalPagesNumber`
* `io.datastorage.LastCheckpointFsyncDuration`

Planned checkpoints run according to `DataStorageConfiguration.checkpointFrequency`.
Dirty-page pressure, checkpoint-buffer pressure, and some administrative operations can trigger an earlier checkpoint.
Successive `LastCheckpointStart` values let you estimate the real interval.
A single early checkpoint does not prove a problem.
A regularly shortening interval together with high `DirtyPages`, high `UsedCheckpointBufferSize`, or growing `TotalThrottlingTime` indicates storage or write pressure.

Approximate checkpoint page-write throughput in MiB/s:

[source,text]
----

> LastCheckpointTotalPagesNumber

Reviewer 2: plz also show "ignite.log" output for checkpoint operations, how it looks

Author: done

Compare several checkpoints and check whether there is enough headroom for the normal write workload.

If headroom is insufficient, use checks and changes that can be verified:

* check storage latency and saturation together with link:monitoring-metrics/new-metrics-system#monitoring-checkpointing-operations[checkpoint metrics];
* use faster production storage devices, as discussed in link:persistence/persistence-tuning#purchase-production-level-ssds[Purchase Production-Level SSDs];
* place data files and WAL on separate physical devices, not just separate directories on the same disk; see link:persistence/persistence-tuning#keep-wals-separately[Keep WALs Separately];
* review link:persistence/persistence-tuning#adjusting-checkpointing-buffer-size[checkpoint buffer] and link:persistence/persistence-tuning#pages-writes-throttling[page write throttling] configuration;
* distribute write load or add server nodes when operational constraints allow it;
* check link:persistence/native-persistence#wal-archive[WAL archive] I/O contention.

JMX checkpoint and data-region metrics are node-local.
Open the same charts or attributes on each server node that owns persistent data.

> .JConsole view of node-local persistence checkpoint metrics

Reviewer 2: ```suggestion
JConsole view of node-local persistence checkpoint metrics
```
is dot expected here ? i miss smth ?

Author: dot is expected due eto ascii doc formatting (.Title for nearest image::)

image::perf-and-troubleshooting/images/healthy-cluster-checkpoint-jconsole.png[JConsole MBeans view showing io/datastorage checkpoint attributes and a persistent dataregion bean]

=== Check Logs and Optional Features

Investigate logs and failure handling for:

* node segmentation;
* repeated system-worker blockage;
* failure-handler activations;
* `OutOfMemoryError` and `IgniteOutOfMemoryException`;
* repeated striped-pool starvation warnings;
* unexpected node restarts.

> Ignite reports critical failures to the configured failure handler; depending on the handler, the result can be node invalidation, failover handling, node stop, or JVM termination.

Reviewer 2: Oh ! seems we need to mention failure handler and it\`s configurations policies here ? wdyt ?

Author: done with crosslinking

See link:perf-and-troubleshooting/handling-exceptions[Handling Exceptions] for failure-handler behavior.
No single local command guarantees the absence of split brain.
Monitor membership from all expected monitoring points, and use logs and failure handling to detect segmentation and discovery failures.

CDC is a feature-specific check, not part of a universal cluster health definition.

> If CDC is enabled, monitor `ignite-cdc.log`, WAL gaps, and consumption lag.

Reviewer 2: theme is not covered at all )) may be it\`s ok for this doc but if you have more info - plz append it

Author: appended

* Adjust link:data-rebalancing[data rebalancing settings] to ensure that rebalancing completes faster when your cluster topology changes.

== How to assess cluster health

> Cluster health is a complex thing. Apache Ignite is capable of demonstrating great performance across different scenarios with varying loads. Therefore, in general terms, a healthy cluster is one whose behavior aligns with your expectations. However, there are some universal aspects that apply to all deployments and warrant attention.

Reviewer 2: ```suggestion
Cluster health is a complex set of metrics and states that need to be analyzed together. Apache Ignite is capable of demonstrating great performance across different scenarios with varying loads. Therefore, in general terms, a healthy cluster is one whose behavior aligns with your expectations. However, there are some universal aspects that apply to all deployments and warrant attention.
```

It is important to understand that a healthy cluster may undergo planned topology changes or temporary load spikes.

The key properties are:

* The cluster is in the intended link:monitoring-metrics/cluster-states[state] and serves only the operations allowed by that state.
* Baseline topology, when it is used or managed manually, matches the expected data-bearing server nodes.
* Data remains consistent: link:tools/control-script#verifying-partition-checksums[`idle_verify`] reports no partition conflicts when the cluster is idle.
* Expected nodes are present, and no unexpected repeated node `JOIN`, `LEFT`, or `FAIL` events or node segmentation are reported.
* link:configuring-caches/partition-loss-policy[Lost partitions] are absent.

> * The number and age of link:tools/control-script#transaction-management[long-running transactions] do not keep increasing; active link:data-modeling/data-partitioning#partition-map-exchange[Partition Map Exchanges (PMEs)] complete; link:monitoring-metrics/new-metrics-system#monitoring-rebalancing[rebalancing] finishes; link:monitoring-metrics/new-metrics-system#monitoring-checkpointing-operations[checkpoint-related metrics] and link:monitoring-metrics/new-metrics#thread-pools[executor queue sizes] return to their usual ranges after the link:monitoring-metrics/new-metrics-system#monitoring-topology[topology] has stabilized and the application workload has returned to its expected level.

Reviewer 2: ```suggestion
* The number and timeout of link:tools/control-script#transaction-management[long-running transactions] do not keep increasing; active link:data-modeling/data-partitioning#partition-map-exchange[Partition Map Exchanges (PMEs)] complete; link:monitoring-metrics/new-metrics-system#monitoring-rebalancing[rebalancing] finishes; link:monitoring-metrics/new-metrics-system#monitoring-checkpointing-operations[checkpoint-related metrics] and link:monitoring-metrics/new-metrics#thread-pools[executor queue sizes] return to their usual ranges after the link:monitoring-metrics/new-metrics-system#monitoring-topology[topology] has stabilized and the application workload has returned to its expected level.
```

----

Run `control.(sh|bat) --baseline` repeatedly or monitor topology metrics to confirm that membership is stable when no maintenance is in progress.
In logs, look for repeated node join, left, fail, segmentation, and exchange-worker messages.
Investigate unexpected repeated `JOIN`, `LEFT`, or `FAIL` events, node segmentation, network failures, or a PME that does not finish.
For PME metrics and transaction checks, see <<check-transactions-and-sql-queries,Check Transactions and SQL Queries>>.

=== Verify Partition Consistency

When the cluster is expected to be idle, run:

[source,shell]
----
control.(sh|bat) --cache idle_verify
----

Successful result:

[source,text]
----
The check procedure has finished, no conflicts have been found.
----

The beginning of a conflict result uses this format:

[source,text]
----
The check procedure has failed, conflict partitions has been found: [counterConflicts=1, hashConflicts=0]
Update counter conflicts:
Conflict partition: PartitionKey [grpId=1544803905, grpName=default, partId=5]
----

The command compares partition update counters and partition hashes between primary and backup copies.
Run it only when data updates are stopped.

> If updates are active, the command can report false conflicts because copies are changing while hashes are being calculated.

Reviewer 2: ```suggestion
If updates are active, the command can report false positive conflicts because copies are changing while hashes are being calculated.
```

A `LOST` count greater than zero is not a normal transient rebalance state.

[source,sql]
----
SELECT STATE, COUNT(*) AS PARTITION_COUNT
FROM SYS.PARTITION_STATES
WHERE STATE IN ('MOVING', 'RENTING', 'LOST')
GROUP BY STATE
ORDER BY STATE;
----

It is also possible to track some node-local cache-group metrics, such as:

* LocalNodeMovingPartitionsCount;
* LocalNodeRentingPartitionsCount;
* LocalNodeRentingEntriesCount.

Use the link:monitoring-metrics/system-views#partition_states[PARTITION_STATES] system view to check partition states:

* `OWNING`: the node is the current primary or backup owner.
* `MOVING`: a partition copy is being loaded on the node during rebalance.
* `RENTING`: an old copy is being removed after ownership changes.
* `EVICTED`: the partition is absent on a node that is no longer an owner; this is not an error by itself.
* `LOST`: the partition is unavailable and must not be used; investigate immediately.

[source,sql]
----
SELECT CACHE_GROUP_ID, PARTITION_ID, NODE_ID, STATE, IS_PRIMARY
FROM SYS.PARTITION_STATES
WHERE STATE IN ('MOVING', 'RENTING', 'LOST')
ORDER BY STATE, CACHE_GROUP_ID, PARTITION_ID, NODE_ID;
----

> In steady state, this query usually should not return `MOVING`, `RENTING`, or `LOST` rows.

Reviewer 2: ```suggestion
In stable cluster state, this query usually should not return `MOVING`, `RENTING`, or `LOST` rows.
```

[#check-checkpoint-pressure]
=== Check Checkpoint Pressure When Persistence Is Enabled

This check applies only to data regions with Native Persistence enabled.
A pure in-memory cluster does not perform persistence checkpoints for its in-memory regions.

A link:persistence/native-persistence#checkpointing[checkpoint] writes dirty pages from RAM to partition files.
Checkpointing itself is a normal background operation.
The problem starts when the application write rate exceeds the effective storage write speed.
Under checkpoint-buffer or dirty-page pressure, Ignite can throttle update threads.
If the checkpoint buffer is exhausted, update processing can stop until the checkpoint completes.

Monitor these metrics for persistent data regions and data storage:

* `io.dataregion.<region>.DirtyPages`
* `io.dataregion.<region>.CheckpointBufferSize`
* `io.dataregion.<region>.UsedCheckpointBufferSize`
* `io.dataregion.<region>.TotalThrottlingTime`
* `io.datastorage.LastCheckpointStart`
* `io.datastorage.LastCheckpointDuration`
* `io.datastorage.LastCheckpointPagesWriteDuration`
* `io.datastorage.LastCheckpointTotalPagesNumber`
* `io.datastorage.LastCheckpointFsyncDuration`

> It looks like this in ignite.log

Reviewer 2: ```suggestion
Corresponding records in <SOME_formatting ? bold ? italic ?>ignite.log looks like as follows:
```

)
ORDER BY NAME;
----

Run `control.(sh|bat) --baseline` repeatedly or monitor topology metrics to confirm that membership is stable when no maintenance is in progress.
In logs, look for repeated `JOIN`, `LEFT`, `FAIL`, or segmentation events and exchange-worker messages.
Investigate unexpected repeated `JOIN`, `LEFT`, or `FAIL` events, node segmentation, network failures, or a PME that does not finish.
For PME metrics and transaction checks, see <<check-transactions-and-sql-queries,Check Transactions and SQL Queries>>.

=== Verify Partition Consistency

When the cluster is expected to be idle, run:

[source,shell]
----
control.(sh|bat) --cache idle_verify
----

Successful result:

[source,text]
----
The check procedure has finished, no conflicts have been found.
----

The beginning of a conflict result uses this format:

[source,text]
----
The check procedure has failed, conflict partitions has been found: [counterConflicts=1, hashConflicts=0]
Update counter conflicts:
Conflict partition: PartitionKey [grpId=1544803905, grpName=default, partId=5]
----

The command compares partition update counters and partition hashes between primary and backup copies.

> Run it only when data updates are stopped.

Reviewer 2: ```suggestion
Run it only on idle cluster i.e. without active data updates.
```
